### ISSUE TITLE
Psionic Optim: add reusable SGD, Adam, AdamW, LARS, and LAMB primitives for Rust-native trainer loops

### ISSUE THREAD

**Issue by AtlantisPleb** [OpenAgentsInc/openagents#3603]
## Why Now

Tinygrad's framework completeness is not only about backward passes. It also has
real reusable optimizer surfaces such as:

- SGD
- Adam
- AdamW
- LARS
- LAMB

Psionic's current train issues talk about optimizer state ownership, but they do
not yet instantiate optimizer parity as a reusable framework-core issue. If the
optimizer logic only appears inside one trainer loop, Psionic still falls short
of Tinygrad-class ML framework completeness.

## Current Baseline

- `#3564` covers the integrated Rust-native training core.
- `#3586` covers distributed optimizer, precision, and memory-sharding
  contracts.
- There is still no dedicated issue for a reusable optimizer library that can be
  used by trainer loops, research loops, and smaller local fine-tune programs.

## Scope

Build the reusable optimizer layer for Psionic.

This issue should cover:

- optimizer traits or contracts over canonical Psionic tensors and gradients
- reusable implementations for at least:
  - SGD
  - Adam
  - AdamW
  - LARS
  - LAMB
- optimizer-state ownership and serialization posture
- step application semantics and update typing
- enough device-awareness that optimizer buffers can be placed honestly rather
  than treated as CPU-only hidden state
- compatibility with later fused update or distributed optimizer work without
  hard-coding the first implementation around one local trainer

## Deliverables

- Reusable optimizer contracts in the correct owner crates.
- Initial optimizer implementations with typed state ownership.
- Tests proving at least one small model can advance through multiple optimizer
  steps without ad hoc trainer-local update logic.
- Doc updates if the optimizer layer changes the current train architecture.

## Acceptance Criteria

- Psionic exposes reusable optimizer surfaces rather than only trainer-private
  update code.
- Optimizer state is inspectable, serializable where appropriate, and compatible
  with later checkpoint lineage.
- At least SGD and Adam are proven end to end; the remaining parity families are
  either landed or have explicit typed hooks that keep the issue honest until
  closed.
- The optimizer layer composes cleanly with `#3564` and `#3586` rather than
  duplicating their future distributed concerns.

## Dependencies

- #3601
- #3602

## Design Constraints

- Keep optimizer logic in reusable Psionic layers, not app code.
- Do not hide optimizer state inside opaque checkpoint blobs only.
- Preserve room for later distributed optimizer and precision work rather than
  baking one local-only assumption into the API.

## Related Existing Work

- #3564
- #3586

## References

- `docs/audits/2026-03-14-tinygrad-parity-target-for-psionic-audit.md`
- `crates/psionic/docs/TRAIN_SYSTEM.md`

**Comment by AtlantisPleb** [comment]
Implemented in `2cba8c2bf` on `main`.

Landed:
- reusable `psionic-train::optimizer` contracts with inspectable step reports/errors and typed state init/application for SGD, Adam, AdamW, LARS, and LAMB
- fixed-budget training core now composes with the reusable optimizer surface instead of carrying trainer-private update logic
- distributed-optimizer accounting and model-IO optimizer-state export/import now understand the expanded optimizer families
- `TRAIN_SYSTEM.md`, `FRAMEWORK_CORE_ACCEPTANCE_MATRIX.md`, and `check-psionic-framework-core-acceptance.sh` now reflect that reusable optimizer primitives are landed and that `#3602` is the remaining framework-core gap in this area

Validated:
- `cargo test -p psionic-train --lib model_io::tests::portable_model_bundle_roundtrips_through_safetensors_manifest -- --exact`
- `cargo test -p psionic-train --lib model_io::tests::portable_model_bundle_roundtrips_new_optimizer_family_variants_through_safetensors -- --exact`
- `scripts/release/check-psionic-framework-core-acceptance.sh --only autodiff-optimizer`
